src/hotspot/cpu/x86/peephole_x86_64.cpp

```diff
@@ -235,6 +236,142 @@ bool Peephole::test_may_remove(Block* block, int block_index, PhaseCFG* cfg_, Ph
+    return false;
+  }
+
+  Node* lea_address = lea_derived_oop->in(AddPNode::Address);
+  Node* decode_address = decode->in(1);
+
+  bool is_spill = lea_address != decode_address &&
+                  lea_address->is_SpillCopy() &&
+                  decode_address->is_SpillCopy();
+
+  // If this is a spill, move lea_address and decode_address one node further up to the
+  // grandparents of lea_derived_oop and decode respectively. This lets us look through
+  // the indirection of the spill.
+  MachNode* decode_spill;
+  if (is_spill) {
+    decode_spill = decode_address->as_Mach();
+    decode_address = decode_address->in(1);
+    lea_address = lea_address->in(1);
+  }
+
+  // The leaP* and the decode must have the same parent. If we have a spill, they must have
+  // the same grandparent.
+  if (lea_address != decode_address) {
+    return false;
+  }
+
+  // Ensure the decode only has the leaP*s (with the same (grand)parent) and a MachProj leaf as children.
+  MachProjNode* proj = nullptr;
+  for (DUIterator_Fast imax, i = decode->fast_outs(imax); i < imax; i++) {
+    Node* out = decode->fast_out(i);
+    if (out == lea_derived_oop) {
+      continue;
+    }
+    if (out->is_MachProj() && out->outcnt() == 0) {
+      proj = out->as_MachProj();
+      continue;
+    }
+    if (out->is_Mach()) {
+      MachNode* other_lea = out->as_Mach();
+      if ((other_lea->rule() == leaP32Narrow_rule ||
+           other_lea->rule() == leaP8Narrow_rule ||
+           other_lea->rule() == leaPCompressedOopOffset_rule) &&
+           other_lea->in(AddPNode::Base) == decode &&
+          (is_spill ? other_lea->in(AddPNode::Address)->in(1)
+                    : other_lea->in(AddPNode::Address)) == decode_address) {
+        continue;
+      }
+    }
+    // There is other stuff we do not expect...
+    return false;
+  }
+
+  // Ensure the MachProj is in the same block as the decode and the lea.
+  if (proj == nullptr || !block->contains(proj)) {
+    // This should only fail if we are stressing scheduling.
+    assert(StressGCM, "should be scheduled contiguously otherwise");
+    return false;
+  }
+
+  // We now have verified that the decode is redundant and can be removed with a peephole.
+  // Remove the projection
+  block->find_remove(proj);
+  cfg_->map_node_to_block(proj, nullptr);
+
+  // Rewire the base of all leas currently depending on the decode we are removing.
+  for (DUIterator_Fast imax, i = decode->fast_outs(imax); i < imax; i++) {
+    Node* dependant_lea = decode->fast_out(i);
+    if (dependant_lea->is_Mach() && dependant_lea->as_Mach()->ideal_Opcode() == Op_AddP) {
+      dependant_lea->set_req(AddPNode::Base, decode_address);
+      // This deleted something in the out array, hence adjust i, imax.
+      --i;
+      --imax;
+    }
+  }
+
+  // Remove spill for the decode if the spill node does not have any other uses.
+  if (is_spill && decode_spill->outcnt() == 1 && block->contains(decode_spill)) {
```

Comment:
You can rearrange this to avoid forward declaration of `decode_spill`:

```
  if (is_spill) {
    MachNode* decode_spill = decode->in(1)->as_Mach();
    If (decode_spill->outcnt() == 1 && block->contains(decode_spill)) {
```

Reply:
Or you concern that inputs of `decode` could be modified and you cached it?

Reply:
Thank you for pointing this out. I was not trying to cache it. Reorganized in [da3f007](https://github.com/openjdk/jdk/pull/25471/commits/da3f007cb786b5a3efcdb155abcf73d8dbcee8af)